compiler/rustc_const_eval/src/interpret/validity.rs

```diff
@@ -491,7 +491,27 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
     ) -> InterpResult<'tcx, bool> {
         // Go over all the primitive types
         let ty = value.layout.ty;
-        match ty.kind() {
+        match *ty.kind() {
+            ty::Pat(inner, _) => {
+                let mut value = value.clone();
+                value.layout.ty = inner;
+                // First visit the inner type to report more targetted errors
+                // if the value is already not valid at the inner type.
+                self.visit_value(&value)?;
+                // Then check the extra pattern restrictions.
+                let scalar = self.read_immediate(&value, "initialized scalar value")?;
+                match (*scalar, value.layout.abi) {
+                    (Immediate::Scalar(scalar), Abi::Scalar(s))
+                    | (Immediate::ScalarPair(scalar, _), Abi::ScalarPair(s, _)) => {
+                        self.visit_scalar(scalar, s)?
+                    }
+                    other => span_bug!(
+                        self.ecx.cur_span(),
+                        "invalid abi {other:?} for pattern type {ty:?}"
+                    ),
+                }
```

Comment:
Isn't this duplicating the scalar range check after this `match`?

I assume for now patterns are restricted to things that can be actually represented in the `Scalar` ABI? If `0 | 5 | 10` is accepted as a pattern then we should properly check it here...